#1 Make the code clearer and simpler.

Selvan66 opened the pull request:
-Add file pins.h with variables
-Add files pin_controls

Selvan66 committed:
#1 Make the code clearer and simpler.

-Add file pins.h with variables
-Add files pin_controls

delipl commented on rafal_szyinski/main.ino:
```diff
@@ -0,0 +1,67 @@
+#include "pin_controls.h"
```
Let's write in C++. Move _pin_controls.h_ to _pin_controls.hpp_

delipl commented on rafal_szyinski/main.ino:
```diff
@@ -0,0 +1,67 @@
+			on_off_button_state = reading;
+			if (on_off_button_state == HIGH)
+			{ 
+					turn_on_off();
```
Here just change the state. Do not make any changes with the LEDs here.

delipl commented on rafal_szyinski/main.ino:
```diff
@@ -0,0 +1,67 @@
+	}
+	last_mode_button_state = reading;
+	
+	switch(state)
```
Add here **DISABLE** state and _default_.

Selvan66 replied:
`DISABLED` and `default` will be doing same thing (`turn_off_color()`) so there's no need writing `DISABLED` state

delipl replied:
Okay. I understand. In this case, I approve but in the future, we have to describe every state and handle the errors in `default` and `UNKNOWN` cases.

delipl commented on rafal_szyinski/pin_controls.h:
```diff
@@ -0,0 +1,65 @@
+void change_mode()
+{
+	turn_off_color();
+	switch(state)
```
It is not necessary to use the switch. 
_state_ is an enumerate. 
DISABLED is 0, LEDR is1.
You can just make:
**++state;**
 and remember to reset when is the last state. Try to use the fast if.
 https://www.w3schools.com/cpp/cpp_conditions_shorthand.asp

delipl commented on rafal_szyinski/pin_controls.h:
```diff
@@ -0,0 +1,65 @@
+#ifndef __PIN_CONTROLS_H__
```
Also, let's write in C++ use **#pragma once**. This is a cleaner include guard.

delipl commented on rafal_szyinski/pins.h:
```diff
@@ -0,0 +1,32 @@
+#ifndef __PINS_H__
```
Here also **#pragma once**

delipl commented on rafal_szyinski/pins.h:
```diff
@@ -0,0 +1,32 @@
+
+#include "Arduino.h"
+
+const int on_off_button = 4;
```
It could be _#define_ but let's use C++ ;D
**constexpr** int on_off_button = 4
It is better #define for C++.
Use this in other _const_ variables

Selvan66 replied:
I think it doesn't matter because we assign values and not a calculations so `const` and `constexpr` have same effect here

delipl replied:
It matters because the `constexp` is filled in compilation level. Let check this out:
> `constexpr` essentially states that a variable or function can be resolved at compile time, but is not guaranteed.

https://stackoverflow.com/questions/66176642/difference-between-constexpr-and-define

Selvan66 replied:
Just like `const` if we don't assign calculation (or it better sound: _expressions_). 

 ```cpp
const     double PI1 = 3.141592653589793;
constexpr double PI2 = 3.141592653589793;
``` 
>Both PI1 and PI2 are constant, meaning you can not modify them. However only PI2 is a compile-time constant. It shall be initialized at compile time. PI1 may be initialized at compile time or run time. 
https://stackoverflow.com/questions/13346879/const-vs-constexpr-on-variables

so if we assagin `PI1` with value it will be initialized at compile time just like `constexpr`.

delipl replied:
I would prefer `constexpr` rather than `const` but okay. I will aprove.

delipl commented on rafal_szyinski/pins.h:
```diff
@@ -0,0 +1,32 @@
+const int blue_pin = 5;
+const int green_pin = 3;
+
+enum States
```
add an UNKNOWN state. Just trust me.

delipl commented on rafal_szyinski/pin_controls.h:
```diff
@@ -0,0 +1,65 @@
+
+void turn_on_off()
+{
+	switch(state)
```
Just change state in _main.cpp_

Selvan66 replied:
so you want that one line if statement that change state was in main.cpp for `turn_on_off` and for `change_mode`?

delipl replied:
I think you can do this with one simple `if` statement in `main.cpp`

delipl reviewed (changes requested):
It is not necessary to compile the program with Arduino IDE therefore separate the definitions and the declarations in files **.hpp** and **.cpp** and change _main.ino_ to **main.cpp**.

Other notes are in the comments attached to the code.

Selvan66 committed:
#2 Make the code clearer and simpler.

-change turn_on_off and change_mode to one line if statement
-split to cpp and hpp file
-adding UNKNOWN state

delipl commented on rafal_szyinski/main.cpp:
```diff
@@ -43,7 +43,12 @@ void loop()
 			if (mode_button_state == HIGH)
 			{ 
-					change_mode();
+				turn_off_color();
```
Don't affect here on the circuit.
It would be better to move `turn_off_color()` before `switch` in the 56 line

delipl commented on rafal_szyinski/main.cpp:
```diff
@@ -43,7 +43,12 @@ void loop()
-					change_mode();
+				turn_off_color();
+				state = state ? static_cast<States>(state + 1) : DISABLED;
+				if (state >= UNKNOW)
```
Try to merge somehow these 2 `if` statements.
```
state = state ? static_cast<States>(state + 1) : DISABLED;
if (state >= UNKNOW)
```

delipl commented on rafal_szyinski/pins.hpp:
```diff
@@ -24,9 +22,8 @@ enum States
 	LEDB,
-	DISCO
+	DISCO,
+	UNKNOW
```
Make `UNKNOW` as a default value of `States`. It should be in the first place.

delipl commented on rafal_szyinski/pins.hpp:
```diff
@@ -24,9 +22,8 @@ enum States
-
-#endif // __PINS_H__
\ No newline at end of file
+States state;
```
You can declare the variable in "one line" like this:
```
enum States {
  UNKNOWN,
  DISABLED,
  LEDR,
  ...
} state;  // like this
```

Selvan66 committed:
#2 changes with states

delipl commented on rafal_szyinski/main.cpp:
```diff
@@ -43,17 +43,15 @@ void loop()
-				{
-					state = LEDR;
-				}
+				
```
It looks very complicated. It was my bad to force you to merge these if statements. :D
Let's leave this but I would do this in that way.
```cpp
if (mode_button_state == HIGH and state != DISABLED)
{ 
    state = states + 1 > DISCO ? LEDR : static_cast<States>(state + 1);
}
```
